### DISCUSSION TITLE
How to add multiple auth middlewares?

### DISCUSSION THREAD

**Discussion by robertlagrant** [litestar-org/litestar#1987] [Q&A] [answered]
### Summary

I'd love some documentation on how to chain together authentication middlewares until either one succeeds, and performs the request, or they all fail and the API returns a 401.

My use case is is I'd like to use the JWT middleware to authenticate some requests, that come from users, from a bearer token, but then also a custom middleware to authenticate API requests from an X-API-Key header. I don't want API calls to 401 because the JWT middleware fails.

I don't know if this is possible through configuration, or if I'd need to make an overarching middleware that somehow calls the others in tandem, but I couldn't find any documentation on how to achieve this. Apologies if it exists and I missed it.

## Funding
* If you would like to see an issue prioritized, make a pledge towards it!
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/litestar-org/litestar/issues/1940">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/1940/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/litestar-org/litestar/issues/1940/pledge.svg">
</picture>
</a>

**Comment by provinzkraut** [comment]
> I don't know if this is possible through configuration, or if I'd need to make an overarching middleware that somehow calls the others in tandem

It's going to have to be the latter since middlewares are designed to be completely independent from each other. We do not have anything about a case like this in the docs, but I think it might be a good addition to a recipes section. @litestar-org/maintainers?

**Comment by robertlagrant** [comment]
Thank you, and for the lightning fast response. I'm happy to help contribute a generic recipe if I can get a pointer on how to programmatically call the middlewares.

**Comment by provinzkraut** [comment, marked as answer]
Middlewares already call other middlewares / ASGI apps, since they themselves are ASGI apps as well.

It usually looks something like 

```python
await self.app(scope, send, receive)
```

So if you were to wrap one middleware into another, you'd simply wrap these calls. 

You can take a look at https://github.com/litestar-org/litestar/blob/a288fd49ec50426da146f8e9071163d5b892c335/litestar/middleware/base.py or the middlewares in https://github.com/litestar-org/litestar/tree/a288fd49ec50426da146f8e9071163d5b892c335/litestar/middleware to see how that's usually implemented.

**Comment by robertlagrant** [comment]
Thank you - just to update we are likely to try this soon, so we'll report back on our progress!

**Reply by Murtagy** [reply]
How well did it go @robertlagrant ?
